### DISCUSSION TITLE
Resetting status of items on shutdown

### DISCUSSION THREAD

**Discussion by panos-tr** [bluegroundltd/transactional-outbox#24] [General]
## Context
The transactional outbox provides a `shutdown` method which is called by the hosting application (for example [here](https://github.com/bluegroundltd/hyperion/blob/424a27d09e8b39773fbf60f00aa918b77995fb0d/modules/reservation/reservation-jobs/src/main/kotlin/com/blueground/reservation/jobs/OutboxJobScheduler.kt#L66)) when it is being shut down. Its purpose is to ensure that the outbox mechanism remains stable between restarts of the hosting application.

## Current Implementation
The [current implementation](https://github.com/bluegroundltd/transactional-outbox/blob/cb983f7ffd17e2aba96bc7b32413d62d1229e6c8/core/src/main/kotlin/io/github/bluegroundltd/outbox/TransactionalOutboxImpl.kt#L145) has 4 steps:
- Initially the executor service enters into shutdown mode.
- Then `awaitTermination` is called on the executor which blocks (with a timeout) waiting for all tasks (i.e. active `OutboxItemProcessor`s) to be completed.
- If any tasks remain after `awaitTermination` returns, `shutdownNow` is called which effectively attempts to **immediately** terminate any active task. It then returns a list of the tasks that were awaiting execution.
- Then if any of the tasks returned from the previous step are of type `OutboxItemProcessor`, it retrieves the items contained in them and [resets their state](https://github.com/bluegroundltd/transactional-outbox/blob/cb983f7ffd17e2aba96bc7b32413d62d1229e6c8/core/src/main/kotlin/io/github/bluegroundltd/outbox/TransactionalOutboxImpl.kt#L168-L170) so that they could be queued by the next execution of `monitor`.

### Bug
The last step above is based on the premise of [filtering tasks based on their type](https://github.com/bluegroundltd/transactional-outbox/blob/cb983f7ffd17e2aba96bc7b32413d62d1229e6c8/core/src/main/kotlin/io/github/bluegroundltd/outbox/TransactionalOutboxImpl.kt#L167). Unfortunately, since introducing decorators (#12), there is a good chance that the actual type of `Runnable` is not `OutboxItemProcessor` which means that **none of the tasks will match the filter**.

This is indeed true for both [blue-brain](https://github.com/bluegroundltd/atlas/blob/5af09fd92f996459021f6bed58344984dfbfaef3/modules/admin/src/main/java/com/blueground/admin/config/OutboxConfiguration.kt#L64-L64) and [hyperion](https://github.com/bluegroundltd/hyperion/blob/d90a8e416ce87587399845d97296613e6f4f7490/modules/reservation/reservation-server/src/main/kotlin/com/blueground/reservation/server/config/OutboxConfiguration.kt#L68).

_(There is also another minor issue with the fact that `OutboxItemProcessor` exposes its internal state through `getItem`, but that is irrelevant to the problem and could easily solved by encapsulating the reset process inside it. The problem would still remain though, in that we would need to access the 'decorated' `OutboxItemProcessor` instead of the `Runnable` being executed.)_

### Consequences
The consequence of an `OutboxItem` not being reset is not fatal. There is already a mechanism that will retry items that are stuck in the `RUNNING` state. However, there is a delay for this items to be picked up again defined at the `rerunAfter` field (typically 1 hour). This means that it would probably be picked up and executed at a much later time than expected which depending on the item type might be acceptable or not.

However, with the introduction of ordering on the outbox items, the problem might be exaggerated since if the item is not picked up (by `fetchEligibleItems`), it won't be included in the grouping process and therefore it will be executed out of order.

### Ignore the issue 😁 
The bug has been present for a lot of time and it has not caused any issue. It seems that the `awaitTermination` part is usually good enough for all tasks to complete or at least start (since the docs state that only 'not started' tasks are returned from `shutdownNot`). Obviously, we should remove the buggy code in order not to give the wrong impression, but it seems the impact is rather minimal.

### Batch update the outbox items
Remove the dependency on the `OutboxItemProcessor` for selecting the items that need to be reset. Instead implement a method on `OutboxStore` that directly resets all items that are still in the `RUNNING` state. This has obvious issues with concurrency (e.g. multiple instances of the application running) so we would need to introduce some kind of identifier on the rows that would allow us to select only the ones that are relevant to the executor being shutdown.

### Provide custom interface for representing task processing and supporting decoration
Implement a custom interface (e.g. `OutboxItemProcess`) that extends `Runnable` and is implemented by `OutboxItemProcessor`. Use that interface for the decoration mechanism, i.e. update `OutboxItemProcessorDecorator` as follows:
```kotlin
@FunctionalInterface
interface OutboxItemProcessorDecorator {
  fun decorate(runnable: OutboxItemProcess): OutboxItemProcess
}
```

This would indeed allow for filtering based on the tasks being of `OutboxItemProcess` type and we could define any required operations on the interface itself.

The obvious drawback of this solutions is that since `OutboxItemProcessorDecorator` is a public interface and one that is used in the configuration of the outbox mechanism, it is a **breaking change**. Therefore all applications would need to be updated in order to use the new functionality and actually provide the expected shutdown support.

**Comment by chris-asl** [comment]
Nice catch! Thanks for the excellent explanation 👏 

I prefer the last suggestion, it's feels clearer and safer to me, compared to the 2nd one.

For me it's an acceptable breaking change TBH, only requiring a minor change. 
We have also the [upgrading guide](https://github.com/bluegroundltd/transactional-outbox/blob/main/UPGRADING.md), which we can enhance and also attach it to the release notes, as we did for [v. 2.0](https://github.com/bluegroundltd/transactional-outbox/releases/tag/v2.0.0).

**Reply by panos-tr** [reply]
Totally agree on the third option being preferable.

That being said I think I have found a fourth solution that solves the issue without the need for a breaking change. It comprises introducing an additional component that serves as the host for the actual processor with the sole responsibility of being the entity that is submitted to the executor. (See #27 for details - work in progress.)

It's not as clean as providing a custom processing interface but I actually don't think we need to select one versus the other. I still think we should go forward with the custom interface but we could also utilize this as a stop-gap solution that allows us to move at our own pace.
